src/utils.ts

```diff
@@ -203,74 +326,188 @@ export function chunkByParagraph(
 }
 
 /**
- * Breaks a single paragraph that exceeds the chunk size limit into smaller sub-chunks.
- * Maintains token-based overlap between sub-chunks and accurate character position tracking.
+ * Calculates the exact character position where overlap should start based on token count.
+ *
+ * Employs binary search to find the precise character position in the original text that
```

Comment:
Backing up to the high level thing here -- for character splitting / overlap, do we need binary search (`O(log n)`)? I would have assumed that we would just split the chars and proceed in a linear fashion (`O(n)`)?

Reply:
Regarding the binary search, here is the delta.

<img width="1631" height="493" alt="image" src="https://github.com/user-attachments/assets/ac909d78-5f5e-4d59-9022-22104130a32d" />

I think that the benefit of the previous implementation is based on the fact that it is splitting the text for every position it tests. The cost of splitting every position is mitigated.

Reply:
<img width="391" height="1016" alt="image" src="https://github.com/user-attachments/assets/23d3a890-5744-4b78-bc3f-69fc6ce2eebb" />

Reply:
<img width="386" height="160" alt="image" src="https://github.com/user-attachments/assets/66119e9f-ed0c-45ab-81a7-5cb2d37fc7de" />